prov/verbs: Allow RDMACM to connect using GIDs

sydidelot opened the pull request:
Hi,

I recently wrote an email to the libfabric-users mailing list to ask if there was a way to run the verbs provider without RDMACM. I didn't get any response so far.

I initially supposed that RDMACM required to have an IP address set for every IB interfaces in order to work. I was wrong, and RDMACM can actually deal with GIDs directly for connection establishment.

The patch in this PR allows the Verbs provider to directly connect to the network adapters using the GID. In other words, the patch allows to use Libfabric even if there is no IP address set for the Infiniband interfaces.

There are significant issues of issues IP addresses for connection establishment:
- It requires to set up/maintain IP addresses for every IB interfaces.
- In the context of multirail (multiple local interfaces that belong  to the same network subnet), it requires specific IP routes to prevent an interface to reply for another one. Connection establishment would fail otherwise.

The GID can be accessed  by looking at the field src_addr returned by `fi_info -p verbs -v`.

Example of output:
`src_addr: fi_sockaddr_ib://fe80:0000:0000:0000:248a:0703:003f:1f6a`

The patch also modifies fabtest so anybody can start testing this new feature. A new option -F allows to specify the address format that is use for the source/destination addresses.

After figuring out the GID of interface that will be used for the server, one can run the following commands with fabtest:

Server: 
`fi_msg_bw -s fe80:0000:0000:0000:248a:0703:003f:1f6a -e msg -p verbs -F FI_SOCKADDR_IB`

Client:
`fi_msg_bw -e msg -p verbs -F FI_SOCKADDR_IB fe80:0000:0000:0000:248a:0703:003f:1f6a`

Signed-off-by: Sylvain Didelot <sdidelot@ddn.com>

shefty commented on fabtests/common/shared.c:
```diff
@@ -2850,6 +2851,12 @@ void ft_parse_addr_opts(int op, char *optarg, struct ft_opts *opts)
+		if (strcmp(optarg, "FI_FORMAT_UNSPEC") == 0)
+			opts->address_format = FI_FORMAT_UNSPEC;
+		else if (strcmp(optarg, "FI_SOCKADDR_IB") == 0)
+			opts->address_format = FI_SOCKADDR_IB;
```
We can drop the unspec check, as that's the default.  Please use a case insensitive comparison.  We should add the other common options here as well, since the user does not know that other values get ignored.

sydidelot replied:
Done. I now use `strncasecmp` and I have added common address formats: `fi_sockaddr`, `fi_sockaddr_in`, `fi_sockaddr_in6` and `fi_sockaddr_ib`

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -1007,6 +1012,162 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+		return -FI_EINVAL;
+	}
+	return 0;
+}
```
There are address printing functions available in the common code.  A GID can be converted using inet_ntop as ipv6

sydidelot replied:
I will modify the code to use `ofi_straddr()`.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -1007,6 +1012,162 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+	if (id)
+		rdma_destroy_id(id);
+	return ret;
+}
```
Can the above function be replaced with filling out a struct rdma_addrinfo locally and avoid calling into the librdamcm to fill this out?

sydidelot replied:
@shefty would you please detail a bit your suggestion? I'm not sure to correcly understand your point.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -1007,6 +1012,162 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+	for (int dev = 0; dev < num_devices; dev++) {
+		context = ibv_open_device(devices[dev]);
+
+		ret = ibv_query_device(context, &device_attr);
```
GID addressing should be limited to devices using the IB link layer.  I don't believe iWarp or RoCE will work.

Opening and querying devices is an expensive operation.  The verbs code does this already somewhere.  Obtaining the GIDs should be moved there to avoid duplicating these costs.  (That said, holy cow, is the verbs initialization code complex.)

sydidelot replied:
Thanks. I see that `vrb_have_device()` opens the devices. I'll merge the two functions together.

Initially, I wanted to use the function `rdma_getaddrinfo()` to retrieve a complete list of local RDMA devices. However, during my testing, the function was always returning only 1 single device and the pointer `ai_next` to the next device was always NULL. Not sure if it's a bug in RDMACM or because I'm using `rdma_getaddrinfo()` incorrectly. 
That's the reason why I had to deal with all this verbs initialization code :-/.

shefty commented on src/common.c:
```diff
@@ -315,7 +320,26 @@ const char *ofi_straddr(char *buf, size_t *len,
+		size = snprintf(buf, *len, "fi_sockaddr_ib://%s", str);
+#else
+		return NULL;
+#endif
```
Treat GID as ipv6 (use inet_ntop) and avoid the #if and adding verbs provider specific code into the common core.

rajachan replied:
@sydidelot  Check out how we implement FI_ADDR_EFA, which does exactly that. EFA does not support IP addressing and directly works witih device GIDs.

sydidelot replied:
@shefty @rajachan Thanks for your help! I checked out the EFA provider and I see that EFA declares its own address format that contains the GID. I can modify the code so an IB address corresponds to a GID and use `inet_ntop` to print it.
If we go that direction, I would suggest renaming `FI_SOCKADDR_IB` because technically, the address won't be a IB `sockaddr` (`struct sockaddr_ib`) address anymore. Do you guys have any thoughts on it?

rajachan commented on fabtests/benchmarks/msg_bw.c:
```diff
@@ -107,6 +107,7 @@ int main(int argc, char **argv)
 	hints->domain_attr->resource_mgmt = FI_RM_ENABLED;
 	hints->domain_attr->mr_mode = opts.mr_mode;
 	hints->domain_attr->threading = FI_THREAD_DOMAIN;
+	hints->addr_format = opts.address_format;
```
Can you break out all the fabtests changes into a separate commit.

sydidelot replied:
Sure, will do.

sydidelot commented on prov/verbs/src/verbs_info.c:
```diff
@@ -1007,6 +1012,162 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+	}
+
+	*dev_name = strdup(ibv_get_device_name(id->verbs->device));
+	if (!(!dev_name)) {
```
Must be `if(!dev_name)`. I will change that

shefty commented:
We need to continue to use struct sockaddr_ib.  That's a commonly defined structure that contains the GID in sib_addr (struct ib_addr), plus other needed fields, like the pkey and scope_id.  A GID by IBTA spec definition is an ipv6 address (except that in reality it's not), but it's safe to use the inet functions to process it.

I was suggesting we could construct the sockaddr_ib ourselves without calling into the librdmacm functions.  As it is, we're letting the rdmacm pick the other fields in the sockaddr_ib without help.  A GID could be associated with multiple pkeys, and despite G in GID meaning 'global', two devices could be assigned the same GID if they are on separate subnets.

sydidelot commented:
@shefty, your comment raises several questions/concerns to me. 

The `sockaddr_ib` strcuture is defined as below: 
```struct sockaddr_ib {                                           
    unsigned short int  sib_family; /* AF_IB */                
    __be16          sib_pkey;                                  
    __be32          sib_flowinfo;                              
    struct ib_addr      sib_addr;                              
    __be64          sib_sid;                                   
    __be64          sib_sid_mask;                              
    __u64           sib_scope_id;                              
};                                                             
```

`sib_addr` can be set using `ibv_query_gid()` and `sib_pkey` with `ibv_query_pkey()`. I'm not exactly sure how to retrieve the other members of the structure without librdmacm.

Regarding your statement below:
> A GID could be associated with multiple pkeys, and despite G in GID meaning 'global', two devices could be assigned the same GID if they are on separate subnets.

Does it mean that the GID is not sufficient for connection establishment and it needs to be associated with a pkey?
As for now, the patch assumes that the GID is the only required data for connection establishment. We will need to change this as well.

shefty commented:
To establish a connection, we need all of the fields in sockaddr_ib, except flowinfo.  A pkey is required. The scope_id is used to select a device when two ports have the same GID.  The latter can occur if the ports are on separate subnets, and the SMs assign the same GID.  Honestly, I think that will be a very rare occurrence, but it's just an index.  The SID is the equivalent to the tcp port number.  That is something the rdmacm knows how to fill in.  I guess we could copy that formatting code.  It just writes a 16-bit port number into a 64-bit value, with the other bits fixed based on the port space.  There are requirements or changes on the CM private data when native IB addressing is used, but I don't remember what they are.

sydidelot commented:
Guys, I have updated the patch but it doesn't include all the changes you requested so far.
Sorry, this PR was inactive for a while but I'm still very interested in using GIDs to establish connections.
@shefty would you please have a look at the function `vrb_get_rdma_addrinfo()` to verify if the construction of the input argument `struct rdma_addrinfo` looks correct to you? (I took some inspiration from `rdmacm`). I will address the remaining comments after I'm sure I'm going to the right direction :-). Thank you!

sydidelot commented:
@shefty @rajachan, I have resurrected the patch with a new version that includes all your review comments. The new format for the `fi_sockaddr_ib` is now the following:
```
fi_sockaddr_ib://[guid]:pkey:port_space:scope_id
```
For example:
```
fi_sockaddr_ib://[fe80::248a:703:1c:dc0c]:ffff:13f:0
```

Please let me know what you think of this new version. And thanks for your time and your help to get this feature approved in Libfabric!

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -153,6 +153,73 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
+	             gid,
+	             &pkey,
+	             &ps,
+	             &scope_id);
```
Please don't one parameter per line.  That just forces developers to scroll more to read the function, making it harder to see the entire picture of what's happening.

sydidelot replied:
I removed the line breaks.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -153,6 +153,73 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
+	if (service)
+		sscanf(service, "%"SCNu16, &port);
+
+	vrb_set_sid(ps, port, sib);
```
The above code looks like it should be in something like ofi_str_to_sib(), which would be a common function to check and convert a string into sib.

sydidelot replied:
Thanks, I have changed the code accordingly.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+{
+	return !(gid->raw[8]  | gid->raw[9]  | gid->raw[10] | gid->raw[11] |
+	         gid->raw[12] | gid->raw[13] | gid->raw[14] | gid->raw[15]);
+}
```
This function is not checking for null gid.  It's checking that only half the git is 0

sydidelot replied:
Please see my comment below.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+	if (ret)
+		return -FI_EINVAL;
+
+	if (vrb_gid_is_null(&gid))
```
What is this check for?

sydidelot replied:
On the system I'm using for development (MOFED stack + IB EDR network adapters), the port attribute `gid_tbl_len` returns multiple GIDs entries, including those that are not set. We need to filter those entries out in order to avoid creating incorrect interfaces.

The function `vrb_gid_is_null()` is inspired by the function `null_gid()` from the utility `ibv_devinfo`:
https://github.com/linux-rdma/rdma-core/blob/master/libibverbs/examples/devinfo.c#L50

Actually, I have changed that function to make it more readable.
```static int vrb_gid_is_null(const union ibv_gid *gid)      
{                                                         
    union ibv_gid null_gid;                               
    memset(&null_gid, 0, sizeof(union ibv_gid));          
    /* Default gid prefix is 0xfe80 */                    
    null_gid.raw[0] = 0xfe;                               
    null_gid.raw[1] = 0x80;                               
                                                          
    return !memcmp(gid, &null_gid, sizeof(union ibv_gid));
}                                                         
```
Please tell me which version for the function do you prefer.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+	(*rai)->ai_family = AF_IB;
+	(*rai)->ai_port_space = RDMA_PS_IB;
+
+	struct sockaddr_ib *sib = calloc(1, sizeof(struct sockaddr_ib));
```
Check for failure.  Please declare all variables at the top of the function (to make it easier to find later).

sydidelot replied:
Done

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+
+        if (ps)
+                sib->sib_sid_mask = htonll(RDMA_IB_IP_PS_MASK);
+        if (port)
```
I think we can assert that ps is set if port is given.

sydidelot replied:
Done.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+}
+
+static int vrb_get_rdma_addrinfo(struct ibv_context *context,
+			     uint8_t port_num, int tbl,
```
tbl should be index, but see below.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
+	if (vrb_gid_is_null(&gid))
+		return -FI_EINVAL;
+
+	ret = ibv_query_pkey(context, port_num, tbl, &pkey);
```
The gid index and pkey index are independent.

sydidelot replied:
Thanks for catching that bug! In case multiple pkeys/gids are set for a port, do we want to create a Libfabric interface for every possible combination of pkeys/gids, or do we want to use the first pair of pkey/gid that is set?

sydidelot commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -153,6 +153,73 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
 	return ret;
 }
 
+static int vrb_get_sin_rai(const char *node, const char *service, uint64_t flags,
```
Tanks. I renamed it to `vrb_get_sib_rai()`

sydidelot commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -153,6 +153,73 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
+	int ret;
+
+	memset(gid, 0, sizeof(gid));
+	sib = calloc(1, sizeof(struct sockaddr_ib));
```
Done

sydidelot commented:
> Please break the first patch up into multiple patches. E.g. changes to formatting the sockaddr_ib string, converting the string to sib, changes to selecting the port space, getting the gids during device init, etc.
@shefty Thanks for the review! I have addressed all of your comments and split up the changes into multiple patches as requested.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -994,6 +994,119 @@ int vrb_get_port_space(const struct fi_info *info)
+
+static int vrb_get_rdma_addrinfo(struct ibv_context *context,
+			     uint8_t port_num, union ibv_gid *gid,
+			     uint16_t pkey, struct rdma_addrinfo **rai)
```
I would rename this to vrb_alloc_ib_addrinfo, and just return struct rdma_addinfo *.  Though the return value is less important to me.

sydidelot replied:
Fixed.

shefty commented on prov/verbs/src/fi_verbs.h:
```diff
@@ -756,6 +756,8 @@ int vrb_ep_destroy_xrc_qp(struct vrb_xrc_ep *ep);
 int vrb_xrc_close_srq(struct vrb_srq_ep *srq_ep);
 int vrb_sockaddr_len(struct sockaddr *addr);
+void vrb_set_sid(enum rdma_port_space ps,
+        uint16_t port, struct sockaddr_ib *sib);
```
github needs a way to comment directly on the commit message...  The commit message needs a better description.  I don't think the changes are related to ipoib.  They seem to just generate native IB addresses for all <port git, pkey> pairs in the system, independent of whether ipoib is enabled on that pair or not.  Obviously, the reason for the change is to handle the case where ipoib is not installed, and I would capture that as the motive.

sydidelot replied:
Thanks. I have updated the commit message as you suggested.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -154,6 +154,99 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
+			return -FI_EINVAL;
+
+		if (service) {
+			ret = asprintf(&straddr, "%s%s:%s", has_prefix ? "" : prefix, node, service);
```
limit line lengths to roughly 80 characters

sydidelot replied:
Fixed.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -154,6 +154,99 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
+		if (ret == -1)
+			return -FI_ENOMEM;
+
+
```
nit: extra blank line

sydidelot replied:
Fixed

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -154,6 +154,99 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
+		if (node || hints->src_addr || hints->dest_addr)
+			return vrb_get_sib_rai(node, service, flags, hints, rai);
+		else
+		return vrb_get_rdmacm_rai(node, service, flags, hints, rai);
```
nit: indentation is off.  Though you could drop the else, and the else below, and move the call to vrb_get_rdmacm_rai() after the if statement.

sydidelot replied:
Fixed.

sydidelot commented on fabtests/common/shared.c:
```diff
@@ -2852,6 +2853,16 @@ void ft_parse_addr_opts(int op, char *optarg, struct ft_opts *opts)
+			opts->address_format = FI_SOCKADDR_IB;
+		else if (!strncasecmp("fi_sockaddr", optarg, 11)) /* keep me last */
+			opts->address_format = FI_SOCKADDR;
+		break;
```
I have updated the man page `fabtests/man/fabtests.7.md` in a separate commit.

sydidelot commented on src/common.c:
```diff
@@ -318,7 +320,20 @@ const char *ofi_straddr(char *buf, size_t *len,
 	case FI_SOCKADDR_IB:
-		size = snprintf(buf, *len, "fi_sockaddr_ib://%p", addr);
+		memset(str, 0, sizeof(str));
+		if (!inet_ntop(AF_INET6, ((uint64_t *)addr + 1), str, INET6_ADDRSTRLEN))
```
I have created a structure `ofi_sockaddr_ib` similar to `sockaddr_ib`, which avoids byte offsets.

sydidelot replied:
Unrelated to this PR but `FI_ADDR_IB_UD` also uses a bunch of byte offsets, which should be converted into a known structure as well:
```
case FI_ADDR_IB_UD:                                              
    memset(str, 0, sizeof(str));                                 
    if (!inet_ntop(AF_INET6, addr, str, INET6_ADDRSTRLEN))       
        return NULL;                                             
    size = snprintf(buf, *len, "fi_addr_ib_ud://"                
            "%s" /* GID */ ":%" PRIx32 /* QPN */                 
            "/%" PRIx16 /* LID */ "/%" PRIx16 /* P_Key */        
            "/%" PRIx8 /* SL */,                                 
            str, *((uint32_t *)addr + 4),                        
            *((uint16_t *)addr + 10),                            
            *((uint16_t *)addr + 11),                            
            *((uint8_t *)addr + 26));                            
```

shefty replied:
Yeah, that shouldn't be defined like that.

sydidelot replied:
Ok, I'll fix that code in a separate patch.

sydidelot commented on src/common.c:
```diff
@@ -318,7 +320,20 @@ const char *ofi_straddr(char *buf, size_t *len,
+		size = snprintf(buf, *len, "fi_sockaddr_ib://[%s]" /* GID */
+			     ":%" PRIx16 /* P_Key */
+			     ":%" PRIx16 /* port space */
+			     ":%" PRIx8 /* Scope ID */,
```
I looked at the existing code and I figured out that `FI_ADDR_IB_UD` was using hex and I decided to do the same. Let me know if you want me to change those fields.

shefty replied:
The pkey under /sys/class/blah blah is reported as hex.  So that makes sense.  I'm not sure about the other two.  As long as the output contains '0x' I guess it doesn't matter, since that makes it obvious.

sydidelot replied:
Do you mean that we should prefix all the hex fields with 0x?
Right now the output looks like this:
```fi_sockaddr_ib://[fe80::248a:703:1c:dc0d]:ffff:13f:0```

After converting the port space and scope id to be unsigned, the output would look like: ```fi_sockaddr_ib://[fe80::248a:703:1c:dc0d]:0xffff:319:0```?

Is that correct?

shefty replied:
Yes, I was suggesting that pkey (and other hex values) be written as 0xffff, rather than ffff.  That at least makes it clear that the value is hex.  This doesn't apply to the IP address itself, since the format is well defined.  We're trying to follow something close to the URI format.  I can't find any reference as to how URI handles hex versus decimal values, or if it's even defined.

So, assuming that there is NOT a URI standard for this, my vote is to report hex values with '0x' as the prefix.  This way, it doesn't matter if we use hex or decimal, we can support both, and the user can use whichever format is most convenient.

If this works, we can report port space and scope id either way.  I don't have a preference.  Sometime in the never going to happen future we can even convert those fields into a more readable string, like "tcp" or "udp" for port space, or "host" for scope.

sydidelot replied:
Thanks for your detailed explanation! I modified the patch so the user can specify both, hex values prefixed with '0x' and decimal values.

sydidelot commented on src/common.c:
```diff
@@ -462,6 +477,58 @@ static int ofi_str_to_ib_ud(const char *str, void **addr, size_t *len)
+	uint16_t ps;
+	uint16_t pkey;
+	uint16_t port;
+	uint64_t scope_id;
```
Done. Thanks!

sydidelot commented on prov/verbs/src/verbs_info.c:
```diff
@@ -994,6 +994,119 @@ int vrb_get_port_space(const struct fi_info *info)
+	sib->sib_sid_mask = htonll(RDMA_IB_IP_PS_MASK);
+
+	if (port)
+		sib->sib_sid_mask |= htonll(RDMA_IB_IP_PORT_MASK);
```
Fixed.

sydidelot commented on prov/verbs/src/verbs_info.c:
```diff
@@ -994,6 +994,119 @@ int vrb_get_port_space(const struct fi_info *info)
+	(*rai)->ai_family = AF_IB;
+	(*rai)->ai_port_space = RDMA_PS_IB;
+
+	sib = calloc(1, sizeof(struct sockaddr_ib));
```
Fixed.

sydidelot commented on prov/verbs/src/verbs_info.c:
```diff
@@ -994,6 +994,119 @@ int vrb_get_port_space(const struct fi_info *info)
+	(*rai)->ai_src_len = sizeof(struct sockaddr_ib);
+
+	sib->sib_family = AF_IB;
+	memcpy(&sib->sib_addr.sib_raw, &gid->raw, 16 * sizeof(uint8_t));
```
Fixed.

sydidelot commented on prov/verbs/src/verbs_info.c:
```diff
@@ -994,6 +994,119 @@ int vrb_get_port_space(const struct fi_info *info)
+				 * so we need to filter them out */
+				ret = ibv_query_gid(context, port, gidx, &gid);
+				if (ret || vrb_gid_is_null(&gid))
+					continue;
```
I'm not sure.. On the system I was using for developing this patch, valid entries didn't follow invalid entries, but I don't know if it's always true.

sydidelot replied:
I can change the code to break the loop. Let me know your decision.

shefty replied:
I have no idea.  The man page doesn't provide any useful information about entries being invalid.  I would keep the code as-is.

sydidelot replied:
Ok, I keep the code as-is and mark the conversation as resolved.

sydidelot commented on prov/verbs/src/verbs_info.c:
```diff
@@ -994,6 +994,119 @@ int vrb_get_port_space(const struct fi_info *info)
+					if (ret)
+						continue;
+
+					dev_name = strdup(ibv_get_device_name(context->device));
```
Fixed.

sydidelot commented:
@shefty Thanks for the review! I have updated the patch series with most of the changes you requested. There are still some open questions/points that need to be addressed (please see my comments above).

shefty commented on include/ofi_net.h:
```diff
@@ -125,6 +125,14 @@ int ofi_discard_socket(SOCKET sock, size_t len);
+#endif
+#ifndef RDMA_IB_IP_PORT_MASK
+#define RDMA_IB_IP_PORT_MASK   0x000000000000FFFFULL
+#endif
```
Hmm... I'm not sure.  For struct ofi_sockaddr_ib, that's a copy of an existing structure from ib.h.  Looking on my system, I don't see that it gets installed.  So to make this work, we have to duplicate at least some of the IB definitions.

I can think of 2 options.  To include rdma_cma.h from this file, we would need to update the build checks to look for that file and only include it if found.  The verbs provider build does that, but we would also need that check as part of the core now.  Option 2: provide our own definitions, basically replacing "RDMA" prefix with "OFI".

Overall option 2 doesn't seem that bad, given that we have to have our own definition of sockaddr_ib.

sydidelot replied:
Thanks! Option 2 seems good to me (and easier to implement :)). I'll update the code to provide our own definitions.

sydidelot replied:
The latest patch series includes the requested changes.

shefty commented:
Please run fi_getinfo_test against the changes.  The Intel CI is showing failures there.

sydidelot commented:
@shefty The failure in `fi_getinfo_test` was because some functions in the common code were not properly updated to support `AF_IB`. That was causing the `rxm` utility provider to assert.

I have added 2 additional patches on top of the existing patch series:
- one path to fix to the common code where `AF_IB` wasn't correctly handled. 
- another patch for the verbs provider removes some code for `sockaddr` address manipulation in favor to the common ofi functions.

shefty commented on src/common.c:
```diff
@@ -476,6 +489,98 @@ static int ofi_str_to_ib_ud(const char *str, void **addr, size_t *len)
+		port = 0;
+
+	*len = sizeof(struct ofi_sockaddr_ib);
+	*addr = calloc(1, *len);
```
Need to check for calloc failure.

sydidelot replied:
Fixed.

shefty commented on prov/verbs/src/verbs_info.c:
```diff
@@ -995,6 +995,111 @@ int vrb_get_port_space(const struct fi_info *info)
+	return rai;
+}
+
+static int vrb_get_gids(struct dlist_entry *verbs_devs)
```
Function name is misleading.  Maybe vrb_get_sib?  vrb_get_ibaddr?

sydidelot replied:
I changed the function name to your first suggestion: vrb_get_sib

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -154,6 +154,96 @@ vrb_get_rdma_rai(const char *node, const char *service, uint64_t flags,
+	if (hints && hints->addr_format == FI_SOCKADDR_IB) {
+		if (node || hints->src_addr || hints->dest_addr)
+			return vrb_get_sib_rai(node, service, flags, hints, rai);
+	}
```
If hints specify an addr_format of SOCKADDR_IB, but we skip the inner if statement, we fall through to vrb_get_rdmacm_rai() below.  Does that path handle the hints properly?

sydidelot replied:
Thanks. I changed that code and added a else branch to the inner if statement, which returns -FI_EINVAL.

sydidelot replied:
Let me correct my previous message: I just recall why we fall through to `addr_format == FI_SOCKADDR_IBvrb_get_rdmacm_rai()`. 
It's needed in the case we want to retrieve all the local interfaces, i.e., neither node and src/dest addresses are set.
Calling `vrb_get_sib_rai()` when `addr_format == FI_SOCKADDR_IB` is safe and returns all the interfaces that are compatible with AF_INET. I'm actually using this path into my application. As a consequence, I'm removing the -FI_EINVAL I previously added.

shefty replied:
nit: The two if statements can be combined into 1.

shefty commented on prov/verbs/src/fi_verbs.h:
```diff
@@ -114,6 +114,7 @@
 
 #define VERBS_NO_COMP_FLAG	((uint64_t)-1)
 
+#define VRB_CM_REQ_HDR_SIZE (36)
```
I think there's confusion as to what's happening here.

When the RDMA CM protocol is used (as determined by the service ID range), then the IB CM REQ packet must carry the RDMA CM header.  The kernel behavior here differs based on the addressing that is used by the app.  For AF_INET or AF_INET6, the kernel is responsible for formatting the RDMA CM header.  The application level private data follows that header.

For AF_IB, the change is that the kernel no longer has the information that it needs in order to format the RDMA CM header.  The only data that it fill out is the version field, which is the first byte of the RDMA CM header.  This is the 'corruption' that the commit message is referencing.  When using AF_IB, user space should fill out the RDMA CM header and pass that to the kernel.  Any application level private data should still follow that header.

The verbs provider should provide the RDMA CM header, even if AF_IB is used.  The GIDs can go in place of the IP addresses.  The application data should follow that. See this link for the RDMA CM header format:

https://github.com/linux-rdma/rdma-core/blob/master/librdmacm/cma.h#L105

struct vrb_cm_data_hdr would follow that data.

sydidelot replied:
Thanks for the detailed answer, @shefty! I learn a lot from your reviews.

I have modified the patch series to fix the issues with the RDMA CM header.
I created a new structure `vrb_rdma_cm_hdr`, which is identical to to the structure `ib_connect_hdr` in RDMA CM.
Source and destination addresses in the RDMA CM header are set based on the addresses stored in `rdma_cm_id::route.addr.{src_addr,dst_addr}`. Regarding `vrb_rdma_cm_hdr::ip_version` and `vrb_rdma_cm_hdr::port`, I set both values to 0. Is that correct?

shefty replied:
I think ip_version should be 6 (a GID is an IPv6 address, except that it's not), and the port should match the lower 16-bits of the SID, depending on whether you're referring to either the source or destination.

shefty commented on include/windows/osd.h:
```diff
@@ -261,6 +261,7 @@ do						\
 #define strcasecmp _stricmp
 #define snprintf _snprintf
 #define sleep(x) Sleep(x * 1000)
+#define strtok_r strtok_s
```
github doesn't make it easy to tell where a patch falls in the commit stack, but please ensure that this patch comes before strtok_r is used in the code, so we don't break the build in the middle.

sydidelot replied:
Thanks! I checked the patch order and strtok_s is used before it's effectively used in the series.

sydidelot commented:
@shefty I have updated the patch series with all your review comments. I also fixed a (bad) issue where OFI_RDMA_PS_IB was mistakenly set to the wrong value.

sydidelot commented:
@shefty Would you please let me know the failure reported by the Intel CI pipeline? I ran fabtest locally with the verbs and it didn't report any error.

shefty commented:
Intel CI failure is unrelated
```
name:   fi_multinode -C rma -n 3 -p "psm2"
07:15:42    result: Fail
```

shefty commented on prov/verbs/src/verbs_cm.c:
```diff
@@ -123,23 +123,41 @@ vrb_msg_ep_prepare_cm_data(const void *param, size_t param_size,
+{
+	struct vrb_rdma_cm_hdr *rdma_cm_hdr = priv_data;
+
+	rdma_cm_hdr->ip_version = 0;
```
This should be 6 << 4

sydidelot replied:
I created a new patch to fix the port and the ip_version.

shefty commented on prov/verbs/src/verbs_cm.c:
```diff
@@ -123,23 +123,41 @@ vrb_msg_ep_prepare_cm_data(const void *param, size_t param_size,
+	struct vrb_rdma_cm_hdr *rdma_cm_hdr = priv_data;
+
+	rdma_cm_hdr->ip_version = 0;
+	rdma_cm_hdr->port = 0;
```
This is the source port.  Lower 16-bits of src_addr->sib_sid.

sydidelot replied:
Thanks. I created a new patch to fix the port and the ip_version.

shefty replied:
Can you fix the commit message?  I'm not sure 'based on Sean's review' will be particularly useful when reading the git log.  :)

sydidelot replied:
I fixed the commit message. Sorry for the rush.

shefty commented on prov/verbs/src/verbs_cm.c:
```diff
@@ -123,23 +123,42 @@ vrb_msg_ep_prepare_cm_data(const void *param, size_t param_size,
+
+	rdma_cm_hdr->ip_version = 6 << 4; /* IPv6 */
+	rdma_cm_hdr->port =
+		(uint16_t)((struct ofi_sockaddr_ib *)&id->route.addr.src_addr)->sib_sid;
```
Should call ofi_addr_get_port() for this

sydidelot replied:
I initially wanted to use ofi_addr_get_port() when I realized that the function returns the port in host byte order. 
Shouldn't we set the port in network byte order?

shefty replied:
I agree, it should be in network order.  But does this cast actually work?  The SID is in network byte order.

sydidelot replied:
Hm... Let me verify if the cast works.

sydidelot replied:
Ok, you are right and my cast actually doesn't work :)
I have updated the code to set the port with` htons(ofi_addr_get_port())`

shefty commented:
bot:aws:retest

sydidelot commented:
That's good news! Thank you!

swelch commented:
@shefty  - I'm assuming with these changes the intent is not supporting GIDs using the IB Verbs XRC transport at this time?

shefty commented:
This only updates the rdma_cm path.  I think Sylvain, who submitted the patches, is only concerned with RC QPs and MSG EPs.

swelch commented:
> This only updates the rdma_cm path. I think Sylvain, who submitted the patches, is only concerned with RC QPs and MSG EPs.

Thanks, sounds good. I had taken a look at the changes and just wanted to make sure he knew there would be additional work required to extend this to XRC if that transport was desired. Since XRC is intended to be used with RxM, it makes sense to me to not worry about XRC support at this time.

sydidelot commented:
@swelch. I'm sorry for my late reply - I didn't notice there was a discussion on-going in the PR. I am not interested in XRC for the moment, but that might change in the future :)
